daemon/daemon_unix.go

```diff
@@ -1177,14 +1177,41 @@ func setupDaemonRoot(config *config.Config, rootDir string, rootIDs idtools.IDPa
 		}
 	}
 
-	if err := ensureSharedOrSlave(config.Root); err != nil {
-		if err := mount.MakeShared(config.Root); err != nil {
-			logrus.WithError(err).WithField("dir", config.Root).Warn("Could not set daemon root propagation to shared, this is not generally critical but may cause some functionality to not work or fallback to less desirable behavior")
-		}
+	if err := setupDaemonRootPropagation(config); err != nil {
+		logrus.WithError(err).WithField("dir", config.Root).Warn("Error while setting daemon root propagation, this is not generally critical but may cause some functionality to not work or fallback to less desirable behavior")
 	}
 	return nil
 }
 
+func setupDaemonRootPropagation(cfg *config.Config) error {
+	rootParentMount, options, err := getSourceMount(cfg.Root)
+	if err != nil {
+		return errors.Wrap(err, "error getting daemon root's parent mount")
+	}
+	if hasMountinfoOption(options, sharedPropagationOption, slavePropagationOption) {
+		return nil
+	}
+
+	if err := mount.MakeShared(cfg.Root); err != nil {
+		return errors.Wrap(err, "could not setup daemon root propagation to shared")
+	}
+
+	// check the case where this may have already been a mount to itself.
+	// If so then the daemon only performed a remount and should not try to unmount this later.
+	if rootParentMount == cfg.Root {
+		return nil
+	}
+
+	if err := ioutil.WriteFile(getUnmountOnShutdownPath(cfg), nil, 0600); err != nil {
```

Comment:
Given that we're not writing to this file, should we just "create" it, something like;

```go
os.OpenFile(name, os.O_RDONLY|os.O_CREATE, 0600)
```

or too much hassle?

Reply:
We don't need access to the fd, so I figured there's no reason to use `OpenFile`